File: server/csharp/MetaObjects/Config/NeutralConfig.cs
// Port-neutral `.metaobjects/config.json` reading.
//
// Reads the NEUTRAL SUBSET (`schema_version`, `sources`, `libraries`). The file also
// carries TypeScript-owned keys (`pending_in_git`, `confidence_thresholds`,
// `extract`, `migrate`, `scope`); those are IGNORED rather than modeled, so a
// new TS-only key never becomes a four-port change. `scope` in particular is
// entirely out of scope for this reader — see
// docs/superpowers/specs/2026-08-19-cross-port-metadata-sources-design.md §4.
//
// `libraries` (FR-043) is read here — not just by TypeScript — because this port has
// no rung-2 native config surface (unlike Java/Kotlin's pom or Python's
// metaobjects.config.yaml): `.metaobjects/config.json` is the ONLY place a C# adopter
// can opt into a shipped library at all. See docs/superpowers/specs/2026-09-13-fr-043-
// feature-and-nfr-packages-design.md §12 Q4 ("should libraries move to
// .metaobjects/config.json? yes, outright... the JVM/C# ports can follow").
using System.Text.Json;

namespace MetaObjects.Config;

public sealed class NeutralConfig
{
    /// The DEFAULT value of `sources` when the key is absent or empty — never a
    /// requirement, and never assumed to exist by any other code path.
    public const string DefaultMetadataDir = "metaobjects";

    private const string MetaObjectsDir = ".metaobjects";
    private const string ConfigFile = "config.json";
    private const int SupportedSchemaVersion = 1;

    /// Each entry is a raw source spec (e.g. `{"path": "model"}`,
    /// `{"resource": "com/acme/model"}`) — kind interpretation belongs to
    /// SourceResolver, not here.
    public IReadOnlyList<IReadOnlyDictionary<string, string>> Sources { get; }

    /// FR-043 selection tokens (e.g. `["iam", "iam/db"]`) — empty when the key is
    /// absent or an empty array. Validated against
    /// <see cref="MetaObjects.Library.LibrarySources.KnownTokens"/> at read time: an
    /// unrecognised token is a config MISTAKE (a human typed it), not the
    /// programmatic-caller case <c>LibrarySources.Resolve</c> tolerates by skipping —
    /// skipped here, it would resurface as ERR_UNRESOLVED_SUPER against the adopter's
    /// own metadata, the wrong place to send someone looking. Mirrors TypeScript's
    /// `resolveCollection` (ERR_UNKNOWN_LIBRARY).
    public IReadOnlyList<string> Libraries { get; }

    private NeutralConfig(
        IReadOnlyList<IReadOnlyDictionary<string, string>> sources, IReadOnlyList<string> libraries)
    {
        Sources = sources;
        Libraries = libraries;
    }

    /// Returns null when `<configDir>/.metaobjects/config.json` does not exist.
    /// A file that EXISTS but is malformed THROWS — swallowing it would make a
    /// typo'd config behave identically to no config at all, silently loading
    /// from a possibly-stale default directory instead. The exact error code
    /// used here is deliberately NOT part of the cross-port contract (only the
    /// raise-don't-degrade behavior is), so callers should match on message,
    /// not code, for this path.
    public static NeutralConfig? Read(string configDir)
    {
        var path = Path.Combine(configDir, MetaObjectsDir, ConfigFile);
        if (!File.Exists(path)) return null;

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(File.ReadAllText(path));
        }
        catch (Exception e) when (e is JsonException or IOException)
        {
            throw new MetaModelException(
                $"{path} exists but could not be read as JSON: {e.Message}",
                ErrorCode.ERR_MALFORMED_JSON);
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
                throw new MetaModelException($"{path} must contain a JSON object", ErrorCode.ERR_TOP_LEVEL_NOT_OBJECT);

            if (!root.TryGetProperty("schema_version", out var v) ||
                v.ValueKind != JsonValueKind.Number ||
                // Compare as a double, not GetInt32(): a float-looking literal like
                // `1.0` is valid JSON and the other three ports all accept it as
                // equal to 1 — GetInt32() used to throw a raw, uncoded
                // FormatException on it instead. TryGetDouble never throws and a
                // genuinely non-integral value (e.g. `1.5`) still correctly fails
                // the equality check below.
                !v.TryGetDouble(out var schemaVersion) ||
                schemaVersion != SupportedSchemaVersion)
            {
                throw new MetaModelException(
                    $"{path}: unsupported schema_version (expected {SupportedSchemaVersion})",
                    ErrorCode.ERR_BAD_ATTR_VALUE);
            }

            var specs = new List<IReadOnlyDictionary<string, string>>();
            if (root.TryGetProperty("sources", out var srcs))
            {
                // A present `sources` key that is not an array must RAISE, not
                // silently read as "absent" — the latter would fall back to the
                // default directory with no diagnostic, exactly the "typo'd config
                // behaves like no config" failure this class exists to prevent (see
                // the file header). This is the GENERAL rule: a present-but-JSON-
                // null `sources` is just as wrong-typed as a present-but-object
                // `sources` — it is not a special case.
                if (srcs.ValueKind != JsonValueKind.Array)
                    throw new MetaModelException($"{path}: \"sources\" must be an array", ErrorCode.ERR_BAD_ATTR_VALUE);

                foreach (var s in srcs.EnumerateArray())
                {
                    if (s.ValueKind != JsonValueKind.Object)
                        throw new MetaModelException($"{path}: each \"sources\" entry must be an object", ErrorCode.ERR_BAD_ATTR_VALUE);

                    var d = new Dictionary<string, string>();
                    foreach (var p in s.EnumerateObject())
                    {
                        // Every source-spec value (`path`/`resource`/`package`) is a
                        // string — silently stringifying a non-string (the prior
                        // behavior) would let {"path": 123} load a directory
                        // literally named "123" rather than failing loudly on the
                        // typo'd config.
                        if (p.Value.ValueKind != JsonValueKind.String)
                            throw new MetaModelException($"{path}: \"sources\" entry \"{p.Name}\" must be a string", ErrorCode.ERR_BAD_ATTR_VALUE);
                        var value = p.Value.GetString()!;
                        if (string.IsNullOrWhiteSpace(value))
                            throw new MetaModelException($"{path}: \"sources\" entry \"{p.Name}\" must not be empty", ErrorCode.ERR_BAD_ATTR_VALUE);
                        d[p.Name] = value;
                    }

                    if (d.Count != 1)
                        throw new MetaModelException($"{path}: each \"sources\" entry must have exactly one key", ErrorCode.ERR_BAD_ATTR_VALUE);

                    specs.Add(d);
                }
            }

            var libraries = new List<string>();
            if (root.TryGetProperty("libraries", out var libs))
            {
                // Same "present-but-wrong-shaped raises" rule as `sources` above.
                if (libs.ValueKind != JsonValueKind.Array)
                    throw new MetaModelException($"{path}: \"libraries\" must be an array", ErrorCode.ERR_BAD_ATTR_VALUE);

                foreach (var l in libs.EnumerateArray())
                {
                    if (l.ValueKind != JsonValueKind.String)
                        throw new MetaModelException($"{path}: each \"libraries\" entry must be a string", ErrorCode.ERR_BAD_ATTR_VALUE);
                    var token = l.GetString()!;
                    if (string.IsNullOrWhiteSpace(token))
                        throw new MetaModelException($"{path}: each \"libraries\" entry must not be empty", ErrorCode.ERR_BAD_ATTR_VALUE);
                    libraries.Add(token);
                }

                var available = MetaObjects.Library.LibrarySources.KnownTokens();
                var unknown = libraries.Where(t => !available.Contains(t)).ToList();
                if (unknown.Count > 0)
                {
                    throw new MetaModelException(
                        $"{path}: \"libraries\" names unknown librar{(unknown.Count == 1 ? "y" : "ies")} " +
                        $"[{string.Join(", ", unknown)}]; available: [{string.Join(", ", available)}]",
                        ErrorCode.ERR_UNKNOWN_LIBRARY);
                }
            }

            // Unknown top-level keys (including `scope`/`migrate`) are IGNORED by
            // design — see the file header.
            return new NeutralConfig(specs, libraries);
        }
    }
}

File: server/python/src/metaobjects/config/neutral_config.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

from metaobjects.errors import ErrorCode, ParseError

#: The DEFAULT value of `sources` when the key is absent or empty — never a
#: requirement, and never assumed to exist by any other code path.
DEFAULT_METADATA_DIR = "metaobjects"

_METAOBJECTS_DIR = ".metaobjects"
_CONFIG_FILE = "config.json"

#: `/^[a-z0-9][a-z0-9._-]*$/` — mirrors the TS `DependencyName` regex in
#: `sdk/src/dependencies.ts` exactly (DESIGN §3.1).
_DEPENDENCY_NAME_RE = re.compile(r"^[a-z0-9][a-z0-9._-]*$")

#: The one transport key a dependency spec may carry.
_DEPENDENCY_TRANSPORT_KEYS = ("path", "npm", "python")


@dataclass(frozen=True)
class NeutralConfig:
    """The port-neutral subset of `.metaobjects/config.json`."""

    #: Raw source specs, each a single-key mapping (`path` / `resource` / `package`).
    sources: list[dict[str, str]]

    #: Declared metadata dependencies (FR-023) — each dict already validated:
    #: `name`, exactly one transport key (`path` / `npm` / `python`), optional
    #: `dir` only beside `npm`/`python`.
    dependencies: list[dict[str, str]]


def read_neutral_config(config_dir: Path) -> NeutralConfig | None:
    """Read the neutral subset from ``config_dir/.metaobjects/config.json``.

    Returns ``None`` when the file does not exist. A file that EXISTS but is
    malformed raises — swallowing it would make a typo'd config behave
    identically to no config at all, silently resolving a possibly-stale
    default directory with no diagnostic.
    """
    path = config_dir / _METAOBJECTS_DIR / _CONFIG_FILE
    if not path.is_file():
        return None

    try:
        raw = json.loads(path.read_text())
    except (OSError, ValueError) as e:
        raise ParseError(
            f"{path} exists but could not be read as JSON: {e}",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        ) from e

    if not isinstance(raw, dict):
        raise ParseError(
            f"{path} must contain a JSON object",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )

    version = raw.get("schema_version")
    # `isinstance(version, bool)` must be checked FIRST and separately: Python's
    # `bool` is a subclass of `int` and `True == 1`, so a bare `version != 1`
    # check accepts `schema_version: true` — a divergence from Java's
    # `isNumber()` and C#'s `ValueKind != Number`, both of which reject a JSON
    # boolean outright.
    if isinstance(version, bool) or not isinstance(version, (int, float)) or version != 1:
        raise ParseError(
            f"{path}: unsupported schema_version {version!r} (expected 1)",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )

    # `.get("sources", [])` only applies the [] default when the key is ABSENT —
    # a present `sources: null` returns None here, which correctly fails the
    # `isinstance(sources, list)` check below rather than silently reading as
    # "absent" and falling back to the default directory with no diagnostic.
    sources = raw.get("sources", [])
    if not isinstance(sources, list) or not all(
        isinstance(s, dict)
        and len(s) == 1
        # Every source-spec value (`path`/`resource`/`package`) must be a
        # non-empty (after stripping whitespace) string — a bare number/
        # boolean/null would otherwise reach `Path()` downstream and raise an
        # uncaught TypeError instead of this coded error, and an empty or
        # whitespace-only `path` would resolve to the config-holding directory
        # itself rather than failing loudly on the typo'd config.
        and all(isinstance(v, str) and v.strip() for v in s.values())
        for s in sources
    ):
        raise ParseError(
            f"{path}: 'sources' must be an array of single-key objects, each "
            "value a non-empty string",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )

    dependencies_raw = raw.get("dependencies", [])
    if not isinstance(dependencies_raw, list):
        raise ParseError(
            f"{path}: 'dependencies' must be an array",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )
    dependencies = [_validate_dependency_spec(d, path) for d in dependencies_raw]
    names = [d["name"] for d in dependencies]
    if len(set(names)) != len(names):
        raise ParseError(
            f"{path}: 'dependencies' names must be unique",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )

    # Unknown top-level keys are IGNORED by design — see the module docstring.
    return NeutralConfig(sources=[dict(s) for s in sources], dependencies=dependencies)


def _validate_dependency_spec(dep: object, path: Path) -> dict[str, str]:
    """Validate one entry of `dependencies` (DESIGN §3.1) and return it
    normalized. Mirrors the TS `DependencySpecSchema` union in
    `sdk/src/dependencies.ts` exactly: a `name`, exactly one transport key
    (`path` | `npm` | `python`), an optional `dir` that is legal only beside
    `npm`/`python`, and no other keys. `mode` is REMOVED (FR-023 §11.3 —
    superseded by explicit `scope.include`); any `mode` key is now an
    unknown-key error like any other unrecognized key.
    """

    def fail(reason: str) -> ParseError:
        return ParseError(
            f"{path}: invalid 'dependencies' entry ({reason})",
            code=ErrorCode.ERR_COLLECTION_NOT_FOUND,
        )

    if not isinstance(dep, dict):
        raise fail("must be an object")

    name = dep.get("name")
    if not isinstance(name, str) or not _DEPENDENCY_NAME_RE.match(name):
        raise fail("'name' must match ^[a-z0-9][a-z0-9._-]*$")

    transports = [k for k in _DEPENDENCY_TRANSPORT_KEYS if k in dep]
    if len(transports) != 1:
        raise fail("exactly one of 'path' / 'npm' / 'python' is required")
    transport = transports[0]

    transport_value = dep[transport]
    if not isinstance(transport_value, str) or not transport_value.strip():
        raise fail(f"'{transport}' must be a non-empty string")

    allowed_keys = {"name", transport}
    if transport in ("npm", "python"):
        allowed_keys.add("dir")
    extra_keys = set(dep.keys()) - allowed_keys
    if extra_keys:
        raise fail(f"unknown key(s): {sorted(extra_keys)}")

    result: dict[str, str] = {"name": name, transport: transport_value}

    if "dir" in dep:
        dir_value = dep["dir"]
        if not isinstance(dir_value, str) or not dir_value.strip():
            raise fail("'dir' must be a non-empty string")
        result["dir"] = dir_value

    return result
